File: go/nemo_relay/pricing/pricing.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package pricing

import nemo_relay "github.com/NVIDIA/NeMo-Relay/go/nemo_relay"

// Config is the canonical pricing config document.
type Config = nemo_relay.PricingConfig

// SourceConfig is implemented by pricing source config structs.
type SourceConfig = nemo_relay.PricingSourceConfig

// InlineSourceConfig embeds a pricing catalog directly in plugin config.
type InlineSourceConfig = nemo_relay.PricingInlineSourceConfig

// FileSourceConfig loads a pricing catalog from a JSON file path.
type FileSourceConfig = nemo_relay.PricingFileSourceConfig

// Catalog is an inline pricing catalog payload.
type Catalog = nemo_relay.PricingCatalog

// ModelPricing is one model pricing catalog entry.
type ModelPricing = nemo_relay.ModelPricing

// TokenRates expresses model rates per one million tokens.
type TokenRates = nemo_relay.TokenPricingRates

// PromptCacheConfig configures cache-read token accounting for a pricing entry.
type PromptCacheConfig = nemo_relay.PromptCachePricing

// TokenRateTier is one prompt-token threshold tier.
type TokenRateTier = nemo_relay.TokenRateTier

// PromptTokenThresholdRateSchedule selects token rates by prompt token thresholds.
type PromptTokenThresholdRateSchedule = nemo_relay.PromptTokenThresholdRateSchedule

// ComponentSpec wraps pricing config as a top-level pricing component.
type ComponentSpec = nemo_relay.PricingComponentSpec

// ConfigReport is the validation or activation report for a plugin config.
type ConfigReport = nemo_relay.ConfigReport

// PluginKind is the top-level plugin kind used by the pricing component.
const PluginKind = nemo_relay.PricingPluginKind

// Pricing unit constants accepted by pricing catalog entries.
const (
	UnitPerToken   = nemo_relay.PricingUnitPerToken
	UnitPerRequest = nemo_relay.PricingUnitPerRequest
	UnitPerSecond  = nemo_relay.PricingUnitPerSecond
	UnitGPUHour    = nemo_relay.PricingUnitGPUHour
)

// Prompt-cache accounting constants accepted by pricing catalog entries.
const (
	CacheReadIncludedInPromptTokens = nemo_relay.CacheReadAccountingIncludedInPromptTokens
	CacheReadSeparate               = nemo_relay.CacheReadAccountingSeparate
)

// NewConfig returns an empty pricing config.
func NewConfig() Config {
	return nemo_relay.NewPricingConfig()
}

// NewInlineSource returns an inline pricing catalog source.
func NewInlineSource(catalog Catalog) InlineSourceConfig {
	return nemo_relay.NewPricingInlineSource(catalog)
}

// NewFileSource returns a file-backed pricing catalog source.
func NewFileSource(path string) FileSourceConfig {
	return nemo_relay.NewPricingFileSource(path)
}

// NewCatalog returns an inline pricing catalog with version 1.
func NewCatalog(entries ...ModelPricing) Catalog {
	return nemo_relay.NewPricingCatalog(entries...)
}

// NewModelPricing returns a model pricing entry with catalog defaults applied.
func NewModelPricing(provider, modelID string) ModelPricing {
	return nemo_relay.NewModelPricing(provider, modelID)
}

// NewTokenRates returns per-token model rates.
func NewTokenRates(inputPerMillion, outputPerMillion float64) TokenRates {
	return nemo_relay.NewTokenPricingRates(inputPerMillion, outputPerMillion)
}

// NewPromptCacheConfig returns default prompt-cache accounting settings.
func NewPromptCacheConfig() PromptCacheConfig {
	return nemo_relay.NewPromptCachePricing()
}

// NewTokenRateTier returns a rate tier with the provided rates.
func NewTokenRateTier(rates TokenRates) TokenRateTier {
	return nemo_relay.NewTokenRateTier(rates)
}

// NewPromptTokenThresholdRateSchedule returns a full-request threshold schedule.
func NewPromptTokenThresholdRateSchedule(tiers ...TokenRateTier) PromptTokenThresholdRateSchedule {
	return nemo_relay.NewPromptTokenThresholdRateSchedule(tiers...)
}

// NewComponentSpec wraps pricing config as an enabled top-level component.
func NewComponentSpec(config Config) ComponentSpec {
	return nemo_relay.NewPricingComponentSpec(config)
}

// Component converts pricing config directly into the shared plugin shape.
func Component(config Config) nemo_relay.PluginComponentSpec {
	return nemo_relay.PricingComponent(config)
}

// ValidateConfig validates a pricing config document without activating it.
func ValidateConfig(config Config) (ConfigReport, error) {
	return nemo_relay.ValidatePricingConfig(config)
}

File: crates/node/pricing.js
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

'use strict';

const plugin = require('./plugin.js');

const PRICING_PLUGIN_KIND = 'pricing';

/**
 * Create a default pricing component config.
 *
 * @returns {object} The minimal pricing config with no sources.
 */
function defaultConfig() {
  return {
    sources: [],
  };
}

/**
 * Create per-token pricing rates with defaults applied.
 *
 * @param {object} [config={}] - Partial token-rate fields to override.
 * @returns {object} A normalized token pricing rates object.
 */
function tokenRates(config = {}) {
  return {
    input_per_million: 0,
    output_per_million: 0,
    ...config,
  };
}

/**
 * Create prompt-cache accounting settings with defaults applied.
 *
 * @param {object} [config={}] - Partial prompt-cache settings to override.
 * @returns {object} A normalized prompt-cache config object.
 */
function promptCache(config = {}) {
  return {
    read_accounting: 'included_in_prompt_tokens',
    ...config,
  };
}

/**
 * Create one prompt-token threshold rate tier.
 *
 * @param {object} rates - Token rates selected by this tier.
 * @param {object} [config={}] - Optional threshold fields to include.
 * @returns {object} A normalized token rate tier object.
 */
function tokenRateTier(rates, config = {}) {
  return {
    ...config,
    rates,
  };
}

/**
 * Create a prompt-token threshold rate schedule.
 *
 * @param {object[]} [tiers=[]] - Ordered threshold tiers.
 * @param {object} [config={}] - Optional schedule fields to override.
 * @returns {object} A normalized rate schedule object.
 */
function promptTokenThresholdRateSchedule(tiers = [], config = {}) {
  return {
    type: 'prompt_token_threshold',
    applies_to: 'full_request',
    tiers,
    ...config,
  };
}

/**
 * Create one pricing catalog entry with defaults applied.
 *
 * @param {object} config - Required and optional model pricing fields.
 * @returns {object} A normalized model pricing entry.
 */
function catalogEntry(config) {
  const { prompt_cache: promptCacheConfig, ...entryConfig } = config;
  return {
    aliases: [],
    currency: 'USD',
    unit: 'per_token',
    ...entryConfig,
    prompt_cache: promptCache(promptCacheConfig),
  };
}

/**
 * Create an inline pricing catalog payload.
 *
 * @param {object[]} [entries=[]] - Pricing catalog entries.
 * @param {object} [config={}] - Optional catalog fields to override.
 * @returns {object} A normalized pricing catalog object.
 */
function inlineCatalog(entries = [], config = {}) {
  return {
    version: 1,
    entries,
    ...config,
  };
}

/**
 * Create an inline pricing source.
 *
 * @param {object} catalog - Pricing catalog payload.
 * @returns {object} A normalized inline source config.
 */
function inlineSource(catalog) {
  return {
    type: 'inline',
    catalog,
  };
}

/**
 * Create a file-backed pricing source.
 *
 * @param {string} path - JSON catalog file path.
 * @returns {object} A normalized file source config.
 */
function fileSource(path) {
  return {
    type: 'file',
    path,
  };
}

/**
 * Create a pricing config from ordered sources.
 *
 * @param {object[]} [sources=[]] - Pricing sources in precedence order.
 * @returns {object} A normalized pricing config object.
 */
function pricingConfig(sources = []) {
  return {
    sources,
  };
}

/**
 * Wrap pricing config as a top-level plugin component.
 *
 * @param {object} config - Pricing component configuration document.
 * @param {{ enabled?: boolean }} [options={}] - Optional component-level flags.
 * @returns {object} A plugin component spec for the pricing plugin.
 */
function ComponentSpec(config, { enabled = true } = {}) {
  return plugin.ComponentSpec(PRICING_PLUGIN_KIND, config, {
    enabled,
  });
}

/**
 * Validate a pricing config document without activating it.
 *
 * @param {object} config - Pricing component configuration document.
 * @returns {object} A structured validation report with diagnostics.
 */
function validateConfig(config) {
  return plugin.validate({
    version: 1,
    components: [ComponentSpec(config)],
  });
}

module.exports = {
  PRICING_PLUGIN_KIND,
  defaultConfig,
  tokenRates,
  promptCache,
  tokenRateTier,
  promptTokenThresholdRateSchedule,
  catalogEntry,
  inlineCatalog,
  inlineSource,
  fileSource,
  pricingConfig,
  ComponentSpec,
  validateConfig,
};
